port/osx/omrosbacktrace_impl.c

```diff
@@ -0,0 +1,418 @@
+	for (i = 0; i < ret; i++) {
+		if (NULL != heap) {
+			*nextFrame = portLibrary->heap_allocate(portLibrary, heap, sizeof(J9PlatformStackFrame));
+		} else {
+			*nextFrame = portLibrary->mem_allocate_memory(portLibrary, sizeof(J9PlatformStackFrame), OMR_GET_CALLSITE(), OMRMEM_CATEGORY_PORT_LIBRARY);
+		}
+
+		if (NULL == *nextFrame) {
+			if (0 == threadInfo->error) {
+				threadInfo->error = ALLOCATION_FAILURE;
+			}
+
+			break;
+		}
+
+		(*nextFrame)->parent_frame = NULL;
+		(*nextFrame)->symbol = NULL;
+		(*nextFrame)->instruction_pointer = (uintptr_t)addresses[i];
+		(*nextFrame)->stack_pointer = 0;
+		(*nextFrame)->base_pointer = 0;
+
+		nextFrame = &(*nextFrame)->parent_frame;
+
+		/* check to see if we should truncate the stack trace to omit handler frames. The faulting address may
+		 * be one less than the address reported by core dump and backtrace on osx. */
+		if ((NULL != prevFrame)
+			&& (NULL != faultingAddress)
+			&& ((*faultingAddress == addresses[i]) || (*faultingAddress + 1 == addresses[i]))
+		) {
+			/* discard all frames up to this point */
+			junkFrames = threadInfo->callstack;
+			threadInfo->callstack = prevFrame->parent_frame;
+
+			/* break the link so we can free the junk */
+			prevFrame->parent_frame = NULL;
+
+			/* correct the next frame */
+			nextFrame = &threadInfo->callstack->parent_frame;
+			/* record how many frames we discard */
+			discard = i + 1;
+		}
+
+		/* save the frame we've just set up so that we can blank it if necessary */
+		if (NULL == prevFrame) {
+			prevFrame = threadInfo->callstack;
+		} else {
+			prevFrame = prevFrame->parent_frame;
+		}
+	}
+
+	/* if we discarded any frames then free them */
+	while (NULL != junkFrames) {
+		J9PlatformStackFrame *tmp = junkFrames;
+		junkFrames = tmp->parent_frame;
+
+		if (NULL != heap) {
+			portLibrary->heap_free(portLibrary, heap, tmp);
+		} else {
+			portLibrary->mem_free_memory(portLibrary, tmp);
+		}
+	}
+
+	return i - discard;
+}
+
+/**
+ * This function takes a thread structure already populated with a backtrace by omrintrospect_backtrace_thread
+ * and looks up the symbols for the frames. The format of the string generated is:
+ * 		symbol_name (statement_id instruction_pointer [module+offset])
+ * If it isn't possible to determine any of the items in the string then they are omitted. If no heap is specified
+ * then this function will use malloc to allocate the memory necessary for the symbols which must be freed by the caller.
+ *
+ * @param portLibrary a pointer to an initialized port library
+ * @param threadInfo a thread structure populated with a backtrace
+ * @param heap a heap from which to allocate any necessary memory. If NULL malloc is used instead.
+ *
+ * @return the number of frames for which a symbol was constructed.
+ */
+uintptr_t
+omrintrospect_backtrace_symbols_raw(OMRPortLibrary *portLibrary, J9PlatformThread *threadInfo, J9Heap *heap)
+{
+	J9PlatformStackFrame *frame = NULL;
+	int i = 0;
```

Comment:
The type of `i` should be `uintptr_t` to match the return type of this function.